### PR TITLE
ipops: Add PTR query support

### PR DESCRIPTION
**Pull request by xkaraman**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [ ] PR should be backported to stable branches
- [x] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description

This PR aims to add support for PRT queries in ipops module.

Some functions were modified to accommodate and reuse the structures already defined for DNS.

Any feedback is appreciated!

### CODE REVIEW

**Comment by oej** [comment]
The topic contains a spelling error - you mean PTR and not PRT :-)

Please add documentation as well

**Comment by xkaraman** [comment]
Oh good catch! 

Yep, will do!

File: src/modules/ipops/ipops_mod.c [hunk 1]
```diff
@@ -1292,6 +1297,31 @@ static int ki_dns_query(sip_msg_t *msg, str *naptrname, str *pvid)
 	return dns_update_pv(naptrname, pvid);
 }
 
+/**
+ *
```
Comment (oej) [inline src/modules/ipops/ipops_mod.c:1301]: There is room for a description of the function here - that's why we have comments before functions.

File: src/modules/ipops/ipops_mod.c [hunk 2]
```diff
@@ -1292,6 +1297,31 @@ static int ki_dns_query(sip_msg_t *msg, str *naptrname, str *pvid)
...
+	str name;
+
+	if(msg == NULL) {
+		LM_ERR("received null msg\n");
+		return -1;
+	}
+
+	if(fixup_get_svalue(msg, (gparam_t *)ip, &ip_address) < 0) {
+		LM_ERR("cannot get the IP address\n");
+		return -1;
```
Comment (oej) [inline src/modules/ipops/ipops_mod.c:1315]: Just thinking loud - any reason you want different returns in different situations in your scripts? All errors return exactly the same value here.

Comment (xkaraman) [reply src/modules/ipops/ipops_mod.c:1315]: I followed the `w_dns_query` as a base and assumed this is how error reporting is done. How would you like me to change them?

**Comment by xkaraman** [comment]
Docs added regarding `ptr_query`.

File: src/modules/ipops/ipops_mod.c (hunk 1 above)
Comment (xkaraman) [reply src/modules/ipops/ipops_mod.c]: Fixed

File: src/modules/ipops/doc/ipops_admin.xml [hunk 3]
```diff
@@ -883,6 +883,57 @@ if(dns_query("test.com", "xyz"))
 
     </section>
 
+    <section id="ipops.f.ptr_query">
+      <title>
+        <function moreinfo="none">ptr_query(ip, pvid)</function>
+      </title>
+
+      <para>
+		  Store the hostname that correspond to ip
```
Comment (oej) [inline src/modules/ipops/doc/ipops_admin.xml:892]: s/ip/an IP address/

Also document if you support both IPv4 and IPv6 clearly. Thanks!

File: src/modules/ipops/doc/ipops_admin.xml [hunk 4]
```diff
@@ -883,6 +883,57 @@ if(dns_query("test.com", "xyz"))
...
+		  in a config variable $ptrquery(pvid=>hostname).
```
Comment (oej) [inline src/modules/ipops/doc/ipops_admin.xml:893]: You mean a "pseudo variable" not a config variable.

File: src/modules/ipops/doc/ipops_admin.xml [hunk 5]
```diff
@@ -883,6 +883,57 @@ if(dns_query("test.com", "xyz"))
...
+      </para>
+
+      <para>Parameters:</para>
+
+      <itemizedlist>
+        <listitem>
+          <para>
+			  <emphasis>ip</emphasis> - string or pseudo-variable containing the ip.
+			  The resulting IP addresses from DNS query are compared with ipaddr.
```
Comment (oej) [inline src/modules/ipops/doc/ipops_admin.xml:902]: In line 901 you just say "ip" and here "ipaddr". I suggest you write clear text on both cases. Like "containing the IP address" and "compared with the IP address"

File: src/modules/ipops/doc/ipops_admin.xml [hunk 6]
```diff
@@ -883,6 +883,57 @@ if(dns_query("test.com", "xyz"))
...
+          </para>
+        </listitem>
+        <listitem>
+          <para>
+            <emphasis>pvid</emphasis> - container id for script variable.
```
Comment (oej) [inline src/modules/ipops/doc/ipops_admin.xml:907]: I don't think we call a pseudo variable a "container" anywhere. If so, I consider it a bug :-)

File: src/modules/ipops/ipops_mod.c (hunk 2 above)
Comment (oej) [reply src/modules/ipops/ipops_mod.c:1317]: Depends if you think there's any point of separating the various errors from each other in the script of if it doesn't matter - the coder just want to know that it failed, not caring about how. There are several functions in other modules, like TM, that has multiple return values. It's really up to you, just a suggestion that you consider this.

File: src/modules/ipops/ipops_pv.c [hunk 7]
```diff
@@ -418,6 +413,91 @@ int dns_update_pv(str *hostname, str *name)
 	return 1;
 }
 
+/*
+*
```
Comment (oej) [inline src/modules/ipops/ipops_pv.c:417]: Add a comment here

File: src/modules/ipops/ipops_pv.h [hunk 8]
```diff
@@ -24,13 +24,46 @@
...
+#define PV_DNS_ADDR 64
+#define PV_DNS_RECS 32
+#define SR_DNS_PVIDX 1
+
+typedef struct _sr_dns_record
+{
+	int type;
+	char addr[PV_DNS_ADDR];
+} sr_dns_record_t;
+typedef struct _sr_dns_item
```
Comment (oej) [inline src/modules/ipops/ipops_pv.h:36]: Empty line missing

**Review by oej — COMMENTED** [review]
You're doing good. Just a few small nits after reading it again.

File: src/modules/ipops/doc/ipops_admin.xml (hunk 6 above)
Comment (henningw) [reply src/modules/ipops/doc/ipops_admin.xml:907]: Actually the module uses already the term "container" in 3 places. So its probably fine in this regards.

File: src/modules/ipops/doc/ipops_admin.xml (hunk 4 above)
Comment (henningw) [reply src/modules/ipops/doc/ipops_admin.xml:893]: The term "config variable" is also already used in the module, but I agree its probably more clear to spell it explictly.

File: src/modules/ipops/doc/ipops_admin.xml (hunk 6 above)
Comment (xkaraman) [reply src/modules/ipops/doc/ipops_admin.xml:907]: These were mostly taken from dns_query docs and just altered it where it made sense.

File: src/modules/ipops/doc/ipops_admin.xml (hunk 4 above)
Comment (oej) [reply src/modules/ipops/doc/ipops_admin.xml:893]: You are right. But it doesn't make it more correct... :-) 

xkaraman - you decide if you want to improve the docs or keep it at the current state.

File: src/modules/ipops/doc/ipops_admin.xml (hunk 6 above)
Comment (xkaraman) [reply src/modules/ipops/doc/ipops_admin.xml:907]: What is the correct term for the `member/container` inside a pv that has multiple keys to get info from?

File: src/modules/ipops/doc/ipops_admin.xml [hunk 9]
```diff
@@ -883,6 +883,93 @@ if(dns_query("test.com", "xyz"))
...
+        <listitem>
+          <para>
+            <emphasis>ipv4</emphasis> - 1 if IPv4 address
+          </para>
+        </listitem>
+        <listitem>
+          <para>
+            <emphasis>ipv6</emphasis> - 1 if IPv6 address
+          </para>
+        </listitem>
```
Comment (xkaraman) [inline src/modules/ipops/doc/ipops_admin.xml:919]: For these i am not exactly sure their purpose. They were accesible before these commits. Are they for counting IPv4/6 in dns_query or just to check if at least one IPv4/6 is associated with hostname?

File: src/modules/ipops/ipops_pv.c (hunk 7 above)
Comment (xkaraman) [reply src/modules/ipops/ipops_pv.c]: Fixed

File: src/modules/ipops/doc/ipops_admin.xml (hunk 5 above)
Comment (xkaraman) [reply src/modules/ipops/doc/ipops_admin.xml]: Removed altogether. there is no comparison done at all. It must be a leftover from DNS query.

File: src/modules/ipops/doc/ipops_admin.xml (hunk 4 above)
Comment (xkaraman) [reply src/modules/ipops/doc/ipops_admin.xml]: Fixed

File: src/modules/ipops/doc/ipops_admin.xml (hunk 3 above)
Comment (xkaraman) [reply src/modules/ipops/doc/ipops_admin.xml]: Fixed

File: src/modules/ipops/ipops_mod.c (hunk 2 above)
Comment (xkaraman) [reply src/modules/ipops/ipops_mod.c:1317]: Main reason i see for separating errors is for debugging purposes. 

Another thing, that might be good, would be to separate the failure over bad input and successful call with no hostname associated ([getnameinfo](https://man7.org/linux/man-pages/man3/getnameinfo.3.html)). Do you think this makes sense to separate and document it?

Other than that, as you have said the code just want to know it had failed and not how.

File: src/modules/ipops/doc/ipops_admin.xml (hunk 4 above)
Comment (miconda) [reply src/modules/ipops/doc/ipops_admin.xml]: I use both terms config variables and pseudo variables, and I am referring to the same concept. Historically the term pseudo-variables was used from developer's perspective, trying to differentiate them from C-style variables. Over the time, I noticed that config-variables are easier to "understand" what one refers to, specially by those coming new to use Kamailio.

In other words, pseudo-variables tries to define them from technical point of view (i.e., no declaration, no explicit type association, automatically available by loading various modules, some read-only, other read-write). The config-variables indicates where these variables are available/can be used.

File: src/modules/ipops/ipops_mod.c (hunk 2 above)
Comment (oej) [reply src/modules/ipops/ipops_mod.c:1317]: Seems like you have reasons for multiple return values, so go ahead.

**Comment by xkaraman** [comment]
So i've rebased to master and squashed some commits to make it more manageable. I have also add some more in-code comments/doc.

If there are no other suggestions/review, I will merge it soon.

**Comment by xkaraman** [comment]
Thanks! Merged!

**Pull request merged**